libyul/optimiser/KnowledgeBase.cpp

```diff
@@ -34,39 +33,31 @@ using namespace std;
 using namespace solidity;
 using namespace solidity::yul;
 
+KnowledgeBase::KnowledgeBase(map<YulString, AssignedValue> const& _ssaValues):
+	m_valuesAreSSA(true),
+	m_variableValues([_ssaValues](YulString _var) { return util::valueOrNullptr(_ssaValues, _var); })
+{}
+
 bool KnowledgeBase::knownToBeDifferent(YulString _a, YulString _b)
 {
-	// Try to use the simplification rules together with the
-	// current values to turn `sub(_a, _b)` into a nonzero constant.
-	// If that fails, try `eq(_a, _b)`.
-
 	if (optional<u256> difference = differenceIfKnownConstant(_a, _b))
 		return difference != 0;
-
-	Expression expr2 = simplify(FunctionCall{{}, {{}, "eq"_yulstring}, util::make_vector<Expression>(Identifier{{}, _a}, Identifier{{}, _b})});
-	if (holds_alternative<Literal>(expr2))
-		return valueOfLiteral(std::get<Literal>(expr2)) == 0;
-
 	return false;
 }
 
 optional<u256> KnowledgeBase::differenceIfKnownConstant(YulString _a, YulString _b)
 {
-	// Try to use the simplification rules together with the
-	// current values to turn `sub(_a, _b)` into a constant.
-
-	Expression expr1 = simplify(FunctionCall{{}, {{}, "sub"_yulstring}, util::make_vector<Expression>(Identifier{{}, _a}, Identifier{{}, _b})});
-	if (Literal const* value = get_if<Literal>(&expr1))
-		return valueOfLiteral(*value);
-
-	return {};
+	VariableOffset offA = explore(_a);
+	VariableOffset offB = explore(_b);
+	if (offA.reference == offB.reference)
+		return offA.offset - offB.offset;
```

Comment:
Is `offA.offset` guaranteed to be bigger than `offB.offset`?

Reply:
No, but it does not matter, this is all taking the wrapping behaviour into account.

Reply:
I have no idea how, but I'll trust you :D

Reply:
It's just that it returns a difference, but whether or not this difference is positive or negative is not an information we can use. It is a u256 anyway (an unsigned number).

What we use are facts of the form:
 - is the difference equal to zero (i.e. are the values equal)?
 - is the difference larger than 31 and less than 2**256-31 (i.e. do the values denote non-overlapping 32-byte areas in memory)?

Reply:
Ahhh, OK, that makes more sense. Thanks.